File: funcscript/parser/syntax/FuncScriptParser.GetIdentifierList.cs
using FuncScript.Block;
using FuncScript.Funcs.Math;
using FuncScript.Model;
namespace FuncScript.Core
{
    public partial class FuncScriptParser
    {
        record GetIdentifierListResult(List<string> IdenList, ParseNode ParseNode, int NextIndex)
            :ParseResult(ParseNode, NextIndex);

        static GetIdentifierListResult GetIdentifierList(ParseContext context, int index)
        {
            int i = SkipSpace(context, index).NextIndex;

            if (i >= context.Expression.Length || context.Expression[i++] != '(')
                return new GetIdentifierListResult(null, null, index);

            var idenList = new List<string>();
            var parseNodes = new List<ParseNode>();

            i = SkipSpace(context, i).NextIndex;
            var (iden, idenLower, _, nodeIden, i2) = GetIdentifier(context, i, false);

            if (i2 > i)
            {
                parseNodes.Add(nodeIden);
                idenList.Add(iden);
                i = i2;

                i = SkipSpace(context, i).NextIndex;
                while (i < context.Expression.Length)
                {
                    if (context.Expression[i] != ',')
                        break;

                    i++;
                    i = SkipSpace(context, i).NextIndex;
                    (iden,  idenLower, _, nodeIden, i2) = GetIdentifier(context, i, false);

                    if (i2 == i)
                        return new GetIdentifierListResult(null, null, index);

                    parseNodes.Add(nodeIden);
                    idenList.Add(iden);
                    i = i2;
                    i = SkipSpace(context, i).NextIndex;
                }
            }

            if (i >= context.Expression.Length || context.Expression[i++] != ')')
                return new GetIdentifierListResult(null, null, index);

            var parseNode = new ParseNode(ParseNodeType.IdentifierList, index, i - index, parseNodes);
            return new GetIdentifierListResult(idenList, parseNode, i);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetIdentifierList.ts
import { ParseContext, ParseNode, ParseNodeType, ParseResult } from "../FuncScriptParser.Main";
import { GetIdentifier } from "./FuncScriptParser.GetIdentifier";
import { SkipSpace } from "./FuncScriptParser.SkipSpace";

export function GetIdentifierList(context: ParseContext, index: number): ParseResult {
    let i = SkipSpace(context, index).NextIndex;

    if (i >= context.Expression.length || context.Expression[i++] !== '(') {
        return { ParseNode: null, NextIndex: index };
    }

    const parseNodes: ParseNode[] = [];

    i = SkipSpace(context, i).NextIndex;
    const { ParseNode:nodeIden,NextIndex:i2} = GetIdentifier(context, i, false);

    if (i2 > i) {
        parseNodes.push(nodeIden!);
        i = i2;

        i = SkipSpace(context, i).NextIndex;
        while (i < context.Expression.length) {
            if (context.Expression[i] !== ',') break;

            i++;
            i = SkipSpace(context, i).NextIndex;
            const {ParseNode:nodeIdenNext,NextIndex: i3} = GetIdentifier(context, i, false);

            if (i3 === i) {
                return { ParseNode: null, NextIndex: index };
            }

            parseNodes.push(nodeIdenNext!);
            i = i3;
            i = SkipSpace(context, i).NextIndex;
        }
    }

    if (i >= context.Expression.length || context.Expression[i++] !== ')') {
        return { ParseNode: null, NextIndex: index };
    }

    const parseNode = new ParseNode(ParseNodeType.IdentifierList, index, i - index, parseNodes);
    return { ParseNode: parseNode, NextIndex: i };
}